tests/jobs/scala/src/main/scala/SpamHam.scala

```diff
@@ -0,0 +1,378 @@
+    def onStart(): Unit = {
+      // Start the thread that receives data over a connection
+      new Thread("Dummy Source") {
+        override def run(): Unit = {
+          receive()
+        }
+      }.start()
+    }
+
+    def onStop(): Unit = {
+      // There is nothing much to do as the thread calling receive()
+      // is designed to stop by itself isStopped() returns false
+    }
+
+    private def receive(): Unit = {
+      while (!isStopped()) {
+        val p = Random.nextDouble()
+        val (label, sampled): (String, Seq[String]) = if (p > 0.5) {
+          ("spam", SpamHamUtils.weightedSample(spamCounts, sentenceLength))
+        } else {
+          ("ham", SpamHamUtils.weightedSample(hamCounts, sentenceLength))
+        }
+        val message: String = s"$label\t${sampled.mkString(" ")}"
+        store(message)
+        Thread.sleep((1000.toDouble / ratePerSec).toLong)
+      }
+    }
+  }
+}
+
+/**
+  * Application that will use a pre-trained model (loaded from S3 or HDFS) and classify messages on the fly
+  * by subscribing to a Kafka topic and classifying the messages that are published to that topic.
+  *
+  * Usage:
+  * {SPARK} {ARGS} {PATH_TO_ASSEMBLY} [topic] [brokers] [modelUri [hdfs:// or s3n://]] [kerberos [true/false]]
+  * where {SPARK} is either `dcos spark run` or `./bin/spark-submit`
+  * {ARGS} are the configuration arguments to Spark (e.g. `--conf spark.cores.max=8` and Kerberos args)
+  * {PATH_TO_ASSEMBLY} is the location of the snapshot jar containing this application
+  *
+  * Example:
+  * dcos spark run --submit-args="{ARGS} --class SpamHamStreamingClassifier \
+  * http://path-to-bucket/dcos-spark-scala-tests-assembly-0.1-SNAPSHOT.jar \
+  * top1 kafka-0-broker.kafka.autoip.dcos.thisdcos.directory:1025 hdfs:///nb_model false"
+  *
+  */
+object SpamHamStreamingClassifier {
+  def main(args: Array[String]): Unit = {
+    if (args.length != 4) {
+      throw new IllegalArgumentException("USAGE: <topic> <brokers> <modelURI> <Kerberos true/false>")
+    }
+
+    val spark = SparkSession
+      .builder()
+      .appName("Spam/Ham SMS Classifier")
+      .getOrCreate()
+
+    // make a streaming context, load the model
+    val ssc = new StreamingContext(spark.sparkContext, Seconds(2))
+    val Array(topic, brokers, modelUri, kerberized) = args
+    val model = PipelineModel.load(modelUri)
+
+    val props = mutable.Map[String, Object](
+      "bootstrap.servers" -> brokers,
+      "key.deserializer" -> classOf[StringDeserializer],
+      "value.deserializer" -> classOf[StringDeserializer],
+      "group.id" -> "use_a_separate_group_id_for_each_stream",
+      "auto.offset.reset" -> "latest",
+      "enable.auto.commit" -> (false: java.lang.Boolean)
+    )
+
+    if (kerberized == "true") {
+      val kerbProps = Map[String, Object](
+        "sasl.kerberos.service.name" -> "kafka",
+        "security.protocol" -> "SASL_PLAINTEXT",
+        "sasl.mechanism" -> "GSSAPI"
+      )
+
+      props ++= kerbProps
```

Comment:
I'd keep the `Map` immutable and add the two. (just to preach the immutability aspect)